pkg/parser/parser.peg

```diff
@@ -313,6 +313,65 @@ InlineAttributes <- "[" attrs:(GenericAttribute)* "]" {
     return types.NewAttributes(attrs.([]interface{}))								
 } 
 
+// Quoted Text attributes (attributes in front of marked up text).
+//
+// There are two syntaxes, a regular form, and another special short-hand form.
+// They behave quite differently from each other, and quite differently from the format
+// used for attribute lists elsewhere.  Both are *extremely* limited.
+//
+// The regular form consists *solely* of a role.  It is the entire text between
+// the open and closing brackets.  No transforms or validation are done.  If this
+// contains illegal characters, asciidoctor passes it straight-through, leading to
+// invalid context.
+//
+// Note that just to make matters more fun, the handling of these lists is somewhat
+// different for constrained vs unconstrained elements.  With constrained elements,
+// the presence of a comma or named attributes is not tolerated, and the list is
+// passed to the output.  With unconstrained markup, any named attributes are simply
+// ignored (everything after the comma is just discarded).  For the sake of sanity
+// we elect to use the more generous form and just discard everything after and including
+// the first comma.  (Perhaps someday more sanity will prevail, and will let us use
+// real named values, including id=, in this form.)  It appears that asciidoctor will
+// trim trailing whitespace, but not leading.  Another win for consistency! (Or not.)
+//
+// The second form is a short-hand that consists of either .role or #id elements,
+// concatenated together with no whitespace permitted between them, or between the
+// enclosing brackets at all.  Anything except a closing bracket (]), dot (.), or hash (#)
+// may be present in these.  Multiple roles may be present, and if so they will be
+// joined with a space and placed in the CSS class.  (Asciidoctor does not escape these,
+// making it possible to create invalid or insecure content.)
+//
+// Corollary 1: There can only be role or ID attributes.
+// Corollary 2: The ID cannot contain a hash or a dot.
+// Corollary 3: There can be only one role if it contains either a hash or a dot.
+// Corollary 4: Neither the ID nor the role may contain a closing bracket.
+//
+// Conclusion: Limit oneself to [0-9A-Za-z_-] in roles and IDs for reliable results.
+//
+
+QuotedTextAttrs <- attrs:(QuotedTextRole / QuotedTextShortHand) {
+    return attrs, nil
+}
+
+QuotedTextRoleWord <- [^\]]* {
+    return strings.SplitN(string(c.text), ",", 2)[0], nil
+}
+QuotedTextRole <- "[" ![#.] role:QuotedTextRoleWord "]" {
```

Comment:
So the issue is that I'd need to that to the QuotedTextRoleWord ... which I could do.  I may test it.  At one point the grammar returned the object, and it was awkward to have that in place because of conversion to []interface{} when presented with multiple items.

Having said that, the grammar now probably will support this suggestion.

Reply:
Ah, I don't think it will work.  because it would then fail to match an empty string.  I.e. I wouldn't be able to handle [].  (right now we do, which is probably not terrifically useful, but I prefer to have a grammar that handles the edge cases.)

Note that the grammar permits a "#" or a "." in the middle, only not in the beginning.  Arguably this is another area where asciidoctor is super inconsistent - as it permits short-hand rules like this for block attribute lists, but not quoted text inline attributes.  Go figure.

Anyway, if you'd rather lose that possibility, I can do the move.  I doubt that it makes a vast amount of difference in performance, but I've not tested.  We're still orders of magnitude faster than the alternatives.

Reply:
I've tried to address this in https://github.com/gdamore/libasciidoc/pull/2
Also, as you will see in my PR, the `AttrRole` attribute value can be a `string` or a `[]string`, depending on how many occurrences were found in the document. This means that the role renderer is also able to deal with these 2 types of values.

Reply:
By the way, I realized that paragraphs and other delimited blocks can also have multiple roles. That's something we can address in #602. For now, I like the idea of having a `string` value when there's a single role on an element, and switching to `[]string` when there is more than one. WDYT?

If we want to always handle roles in `[]string`, then I would suggest that we rename the `AttrRole="role"` attribute to `AttrRoles="roles"` to reflect the cardinality. We could do that in #602.

Reply:
I'd be ok with renaming it to Roles.  I'd kind of like this to always be a []string for consistency.  The renderer code does turn it into a single string.

Reply:
yes, that will be better for the sake of consistency. I'll do that in #602 then ;)